### ISSUE TITLE
[Architecture] Change model annotations to @ProviderType

### ISSUE THREAD

**Issue by krakong** [adobe/aem-core-wcm-components#631]
## Feature Request

Currently, the model interfaces are either annotated with `@ConsumerType` or not at all.

O'Reilly describes the difference between OSGi consumer types and provider types as following (from [https://www.oreilly.com/library/view/building-modular-cloud/9781449345143/ch04.html](https://www.oreilly.com/library/view/building-modular-cloud/9781449345143/ch04.html)
> Some interfaces are implemented by, at most, a few implementations but used by many. LogService is an example of this; there are only a handful of libraries that implement the LogService interface, but the interface is used everywhere code needs to log something. We call such interfaces provider types. Other interfaces are used the other way around: they are implemented many times but are only used by a handful of implementations. javax.servlet.Filter is an example of this; many servlet filters are created by developers, but they are only consumed by a handful of application servers. We call such interfaces consumer types.

Based on that description, the core components model interfaces should be of `@ProviderType` because they are usually implemented only once.

This annotation exists only during compile-time and is commonly used for semantic versioning.
It would make sense to either:
 - Remove all occurrences of the `@ConsumerType` annotations since everything is by default considered as a consumer type
- Annotate all models as `@ProviderType`

**Comment by vladbailescu** [comment]
We explicitly marked the models as `@ConsumerType` so that adding new methods would not result in  major version changes for the packages, and to make it clear that it's an intended behaviour. We believed this will allow people that create custom models to not have to change and re-bundle their implementations, so that upgrades to core components are just a drop-in replacement.

This proved to not hold true for all cases, see the discussion in #480. Moving forward, we need to rethink how we handle the case of adding new properties without breaking implementations, but this is probably a discussion best suited for our mailing list (#513).

**Comment by krakong** [comment]
Thanks for the reply!
I agree that adding new methods to models shouldn't result in major version bumps.
But isn't `@ProviderType` the correct annotation for that behavior?
From https://osgi.org/javadoc/r6/annotation/org/osgi/annotation/versioning/ProviderType.html:
> A non-binary compatible change to a provider type normally requires incrementing the **minor version** of the type's package. This change will require all providers to be updated to handle the change, but consumers of that package will not require changes since they only use, and do not implement, the provider type.

**Comment by vladbailescu** [comment]
I think most of the developers that want to customise the models will write implementations (providers) and not consumers.

**Comment by gabrielwalt** [comment]
Closing this as it is working as designed, and there seems to be no consensus on the different approach so far. Please reopen, or create a new issue, if you want to continue on this topic.
